### PR TITLE
add PARSL_MPI_PREFIX as prefix

### PR DESCRIPTION
**Pull request by tomdemeyere**
## Summary of Changes

Attempt to solve https://github.com/Quantum-Accelerators/quacc/issues/2076 If the user provides `parsl_resource_specification` then it takes precedence in the command prefix. This should also be done for ONETEP, VASP, ... but I don't know how happy you are with this whole concept.

Small reminder about why we care about that: This allow Parsl user to use any kind of launcher, mainly mpirun

### Requirements

- Your PR should be focused on a [single feature addition or bugfix](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/best-practices-for-pull-requests#write-small-prs).
- Your PR should have relevant, comprehensive [unit tests](https://quantum-accelerators.github.io/quacc/dev/contributing.html#unit-tests).
- Your PR should be on a custom branch and _not_ be named `main`.

### Notes

- Before contributing for the first time, read the ["Guidelines" section](https://quantum-accelerators.github.io/quacc/dev/contributing.html#guidelines).
- If you are an external contributor, you will see a comment from [@buildbot-princeton](https://github.com/buildbot-princeton). This is solely for the maintainers.
- When your code is ready for review, ping one of the [active maintainers](https://quantum-accelerators.github.io/quacc/about/contributors.html#active-maintainers).

### CODE REVIEW

**Comment by buildbot-princeton** [comment]
Can one of the admins verify this patch?

**Comment by Andrew-S-Rosen** [comment]
Oh, I see. This might be why my half-hearted attempt to use the new `MPIExecutor` did not work as planned (https://github.com/Quantum-Accelerators/quacc/pull/2167).

It would probably make sense to have this taken care of early on, like when the settings are initialized. For instance, [here](https://github.com/Quantum-Accelerators/quacc/blob/b4344c84eee3221276c4bc461e717b7e74f5157e/src/quacc/settings.py#L464) one could swap out the 0th index of the tuple with the environment variable if supplied. This would make it easier to prevent bugs since we won't have to make changes in multiple places. It is a bit annoying to have to do this for every single MPI executable, but I don't really see a way around that given the current Parsl setup, and if it's all confined within a single `settings.py` file, at least it's not too cumbersome to deal with.

File: src/quacc/calculators/espresso/espresso.py [hunk 1]
```diff
@@ -424,9 +424,13 @@ def __init__(
             .get("pseudo_dir", str(SETTINGS.ESPRESSO_PSEUDO))
         )
 
+        prefix = (
+            os.environ.get("PARSL_MPI_PREFIX", "") or SETTINGS.ESPRESSO_PARALLEL_CMD[0]
```
Comment (Andrew-S-Rosen) [inline src/quacc/calculators/espresso/espresso.py:428]: Small nit: You can just do `os.environ.get("PARSL_MPI_PREFIX")` without assigning a default. The default is `None`.

Comment (tomdemeyere) [reply src/quacc/calculators/espresso/espresso.py:428]: Yes, I don't know what happened yesterday evening but I just got confused...

**Comment by tomdemeyere** [comment]
Debugging Parsl scripts is a nightmare, I still didn't find a way to cleanly show stdout somewhere...

This should work, I just wanted to make sure of what happens if you send jobs with different specification inside a subflow, but suddenly I am getting hit with a:

 File "/Users/tomdm/repos/quacc/src/quacc/wflow_tools/decorators.py", line 617, in wrapper
    return func(*f_args, **f_kwargs)
  ^^^^^^^^^^^^^^^^^
  File "/Users/tomdm/repos/quacc/src/quacc/recipes/espresso/core.py", line 87, in static_job
    is_metal = check_is_metal(atoms)
      ^^^^^^^^^^^^^^^^^
NameError: name 'check_is_metal' is not defined

When trying to run a static_job with Espresso on my local machine

``` python

from quacc import SETTINGS

SETTINGS.WORKFLOW_ENGINE = "parsl"

import os

import parsl
from parsl import bash_app, python_app
from parsl.channels import LocalChannel
from parsl.config import Config
from parsl.executors import HighThroughputExecutor
from parsl.providers import LocalProvider
from quacc.recipes.espresso.core import static_job
from ase.build import molecule
from quacc import job, redecorate, subflow

config = Config(
    executors=[
        HighThroughputExecutor(
            label="htex_local",
            cores_per_worker=1,
            max_workers=1,
            provider=LocalProvider(
                channel=LocalChannel(),
                init_blocks=1,
                max_blocks=1,
            ),
        )
    ],
)

resource_spec_1 = {
    "num_nodes": 2,
    "ranks_per_node": 2,
    "num_ranks": 4,
}

resource_spec_2 = {
    "num_nodes": 1,
    "ranks_per_node": 1,
    "num_ranks": 1,
}

parsl.load(config)


atoms = molecule("H2O")

"""
@job
def test_mpi_apps():
    from os import environ

    return [environ.get("PARSL_RANKS_PER_NODE"), environ.get("PARSL_NUM_NODES"), environ.get("PARSL_NUM_RANKS"), environ.get("PARSL_MPI_PREFIX"), environ.get("PARSL_MPIEXEC_PREFIX"), environ.get("PARSL_MPI_NODELIST")]
"""
    

test_mpi_apps_1 = redecorate(static_job, job(parsl_resource_specification=resource_spec_1))
#test_mpi_apps_2 = redecorate(static_job, job(parsl_resource_specification=resource_spec_2))

#@subflow
#def subflow_mpi_apps():
#    return [test_mpi_apps_1(atoms), test_mpi_apps_2(atoms)]

future = test_mpi_apps_1(atoms)

print(future.result())```

If you have time can you tell me what you get from running this script? (nothing to change normally)

**Comment by tomdemeyere** [comment]
Just tried the same script on master and get the same thing, I guess there are still things I don't understand about Parsl!

In any case, it seems to not be related to this MR.

To be completely honest this would require much more testing (with proper HPC etc...) which practically cannot be done here? I would suggest that for now Quacc goes with that and wait for other users feedback (or me but after the PhD, when I have time 😅)

**Comment by Andrew-S-Rosen** [comment]
Thanks! I'm going to go ahead and merge this as-is. When I go ahead and try to solve https://github.com/Quantum-Accelerators/quacc/issues/2116, I'll be testing things out over HPC and with different codes, which will likely spot any problems for me to address. But this PR will be a good point for me to build from. Thanks!

**Pull request merged**